### DISCUSSION TITLE
Restrict the schemas to valid values

### DISCUSSION THREAD

**Discussion by balazsdukai** [cityjson/specs#186] [Ideas]
I experimented with the popular [JSON Schema Faker](https://json-schema-faker.js.org/) tool to mock CityJSON data.
I used the [combined v2.0 schema](https://3d.bk.tudelft.nl/schemas/cityjson/2.0.0/cityjson.min.schema.json) to fake values for the CityJSON members.
It does generate a seemingly schema-valid CityJSON object, however many members have invalid values, for instance negative vertex indices in the boundaries.
The reason for this is that in several cases the schema does not restrict the values sufficiently.

Below is are the issues and potential fixes that I discovered from the generated CityJSON.

- The version number is invalid. It should be restricted to the CityJSON version that is described by the schema.

```json
"version": "3.4"
```

- CityObject family references point to nonexistent objects. This is probably not possible to prescribe solely with JSON Schema.

```json
"CityObjects": {
  "et_": {
    "type": "BuildingInstallation",
    "parents": [
      "ex fugiat",
      "sed veniam ut dolore"
    ],
    "children": [
      "ipsum nostrud"
    ]
  }
}
```

- Geometry LoD has invalid value. The `lod` should be restricted to the allowed values.

```json
"lod": "9"
```

- Vertex indices in the boundaries have negative values. The vertex indices should be restricted to positive integers. [This example](https://json-schema.org/draft/2020-12/json-schema-core#name-schema-re-use-with-defs) shows that this should be possible.

```json
"boundaries": [
  [
    44174456,
    49920625,
    -95552225,
    -13564149
  ]
]
```

- Semantic object does not have `type`. In the example below, the second semantic object does not have a `type` member, which is required according to the specs. 

```json
"semantics": {
  "surfaces": [
    {
      "type": "ClosureSurface",
      "laboris_4": "sit enim in",
      "amet62": true
    },
    {
      "deserunt_af": 18338638,
      "quis_2": -44755250,
      "in_e": 51750499,
      "adb_": "enim irure",
      "velit__": 73779662,
      "tempor_73": false
    }
  ]
}
```

I don't see drawbacks from applying these restrictions where possible, but maybe others see it differently?

**Comment by hugoledoux** [comment]
Great that you start this 👍

1. cityjson version: then we encode manually all versions? needs to be updated as we go though...
2. no way to enforce this indeed AFAIK
3. no negative should be there, but upper-bound cannot be enforced and you'll have vertex indices w/o any vertex object
4. indeed, the type should be mandatory! The specs mention this, but the schema do not have that line.

**Comment by balazsdukai** [comment]
Closing this because discussion was transferred to https://github.com/cityjson/specs/issues/187
